[mkcal] Remove usage of ::Ptr in service handler.

dcaliste opened the pull request:
This is an API break, but there is no sense to use ::Ptr in this API. Implementation should not keep any pointer on the passed notebook and incidences. More over, having the ::Ptr avoids the use of the service in context where only plain notebook and incidences are available.

Rationalise the ServiceHandler API, not passing the ExtendedStorage pointer when it's not used in the actual interface and only there to check notebook validity with respect to the storage.

@pvuorela, this is not in a hurry, but I would like your opinion on the opportunity to update this plugin API. Having it in ::Ptr everywhere and with useless additional arguments makes it difficult to use it in an environment different from the one now : a ExtendedStorage in the same thread as the service. While I would like to use it also in a different thread because none of the function are actually related to an ExtendedStorage object but they are much more local, either using a Notebook or an Incidence.

It will require to update the caldavinvitation plugin and maybe one in EAS (while actually it seems to me that the defaultInvitationPlugin is actually playing this role).

pvuorela commented on plugins/defaultinvitationplugin/defaultinvitationplugin.cpp:
```diff
@@ -318,11 +318,13 @@ QString DefaultInvitationPlugin::pluginName() const
 
 QString DefaultInvitationPlugin::icon() const
 {
+    d->mErrorCode = ServiceInterface::ErrorOk;
```
This error api is strange. The ServiceInterface defines it as "Sometimes the true/false is not enough, so in case of false more details can be obtained".

Could interpret that as only  (some?) bool methods should be setting the error code? Or how would one know when to expect something.

pvuorela replied:
Let's rather limit the scope of the errors reported by the error code rather than expand it. I don't see anything useful coming out of icon getter failing and the code calling handling that depending on the error code.

What could make sense is these "True if OK, false otherwise" methods giving more info in case of failure. One could, though, question whether they should just an error code instead of boolean then. If I'd try to justify current way, I could say this allows code like "if (!operation())  { /* handle error*/ }

dcaliste replied:
Ok, I see. I've thus let it as it is now.

pvuorela commented on src/servicehandler.cpp:
```diff
@@ -194,181 +92,236 @@ ServiceHandler::ServiceHandler()
-        }
+    d->mError = ErrorOk;
+    if (plugin) {
+        d->mError = ErrorOk;
```
Assigning the same value as two lines above.

pvuorela commented on src/servicehandler.cpp:
```diff
@@ -194,181 +92,236 @@ ServiceHandler::ServiceHandler()
-        bool res = service->shareNotebook(notebook, sharedWith);
+    d->mError = ErrorOk;
+    if (plugin) {
+        d->mError = ErrorOk;
```
Redundant.

dcaliste replied:
Removed here also.

pvuorela commented on src/servicehandler.cpp:
```diff
@@ -194,181 +92,236 @@ ServiceHandler::ServiceHandler()
-        QString res = service->displayName(notebook);
+    d->mError = ErrorOk;
+    if (plugin) {
+        d->mError = ErrorOk;
```
...

dcaliste replied:
Same same, thanks.

pvuorela commented on src/servicehandler.cpp:
```diff
@@ -194,181 +92,236 @@ ServiceHandler::ServiceHandler()
+            // service needed to get possible error, because
+            // invitationhandlerinterface doesn't have error-function
+            // This is utterly broken by design, except when the invitation plugin is also a service plugin
+            ServiceInterface *s = service(notebook.pluginName());
```
Could prefer to avoid single character variables unless they are loop iterators or such.

dcaliste replied:
Ah sure, I've put a more sensible variable name.

pvuorela reviewed (commented):
Partly again pondering how much we gain by avoiding the extra Ptr, but at least the service handler usage is lot more limited.

dcaliste commented:
I've rebased this PR also. Maybe dropping the ::Ptr stuff from the API is not worth the API breakage, but working now on the replacement, or improvement, of Extended{Calendar,Storage} from nemo-qml-plugin-calendar to something that would allow incidence sharing UID in multi notebooks, I think, it may be worth to remove the Extended{Calendar,Storage} from this service API, since it's not really needed and can be better replaced by the Notebook directly.

pvuorela reviewed (commented):
Let's keep the Ptr for now. It's common convention on kf5-calendarcore and shouldn't matter on the calls too much. Less adjustments to make on depending plugins and keeps the other changes more visible if not changing the types.

dcaliste commented:
Ok, I've put back the ::Ptr where they were. It implies less changes to the downstream code. But it's a bit ugly to have it here where it's not necessary and sometime even harmfull since we need to pass a ::Ptr-ified pointer.

pvuorela commented on src/notebook.h:
```diff
@@ -424,6 +424,12 @@ class MKCAL_EXPORT Notebook
+       Provide the email address associated to the notebook, if any.
+       The address is retrieved via the plugin mechanism, see setPluginName().
+    */
+    QString emailAddress() const;
```
Why do we need or want this here when it's already and still available on ServiceHandler? The Notebook class was earlier just a data container without references to other classes while this adds references outside of it. Also there's account on the Notebook too already.

dcaliste replied:
I'm proposing this extension, because this `emailAddress` case is like a property of a Notebook. Accounts have various way to set-up the notebook email address, and thus provide via the ServiceHandler plugin a way to get this property.

For me, it looks more intentional to write `email = notebook->emailAdress();` than `email = mKCal::ServiceHandler::instance()->emailAdress(notebook);`. `sendResponse()` for instance is more like an action that is handled by a plugin, and calling it with `mKCal::ServiceHandler::instance()->sendResponse()` seems logical as an action of the service handler.

I can remove this addition if you prefer, it's not even related to the matter at hand : the removal of Extended*.

pvuorela replied:
I can understand the motivation, but let's skip this for now. The helper method is not much code and the need for it is not too widely spread.

dcaliste replied:
Ok, I've removed it from the commit and updated the nemo-qml-plugin-calendar PR accordingly.

pvuorela commented on src/notebook.cpp:
```diff
@@ -518,3 +519,13 @@ QList<QByteArray> Notebook::customPropertyKeys() const
+QString Notebook::emailAddress() const
+{
+    if (!d->mPluginName.isEmpty()) {
+        Ptr ptr(new Notebook(*this));
```
The notebook is more important detail than being a pointer or not, could use name more towards that.

pvuorela commented on src/servicehandler.cpp:
```diff
@@ -194,122 +92,171 @@ ServiceHandler::ServiceHandler()
+        QString res = plugin->emailAddress(notebook);
         if (res.isNull()) {
-            d->mError = (ServiceHandler::ErrorCode) service->error(); //Right now convert directly
+            d->mError = (ServiceHandler::ErrorCode) plugin->error(); //Right now convert directly
```
Was already like this, but the null QString is not defined to be any indication of error, and just returning a QString() is quite conventional way to return an empty QString. Moreover as mentioned we don't need error handlings for plain property getters so let's remove this and the mError assignments a few lines above.

Applies to couple other similar cases around too.

dcaliste replied:
Ok, I see. I've removed the error handling for `icon`, `emailAddress` and `displayName` getters.

pvuorela commented on src/servicehandler.h:
```diff
@@ -76,75 +74,60 @@ class MKCAL_EXPORT ServiceHandler : QObject
     /** Send the invitation to the list of people stated as attendees.
       It would load the appropriate plugin to do it, and if there
```
Bonus points for removing these repeating "it would load..." lines which are not even good English :)

dcaliste replied:
Sure, I've removed these lines and added a brief comment in the ServiceHandler class definition mentioning that plugins are used based on the `Notebook::pluginName()` return.

pvuorela commented on src/servicehandler.cpp:
```diff
@@ -194,122 +92,171 @@ ServiceHandler::ServiceHandler()
+int ServiceHandler::downloadAttachment(const Notebook::Ptr &notebook, const QString &uri, const QString &path)
 {
-    if (storage.isNull() || notebook.isNull())
+    if (notebook->isRunTimeOnly()) {
```
Was already checking this, but wonder if needed. A runtime notebook could in theory do some downloading? At the moment not used, of course. Same thing on deletion side.

dcaliste replied:
It was there already in the `executePlugin()` routine. I removed this routine, but kept the chek (I've removed the readOnly() one considering that it's a hint only).

Do you prefer to remove this test ?

dcaliste replied:
Ah, no sorry. I mixed up with the sendResponse() case and friends. I've removed the test here and in delete.

pvuorela commented:
> Ok, I've put back the ::Ptr where they were. It implies less changes to the downstream code. But it's a bit ugly to have it here where it's not necessary and sometime even harmfull since we need to pass a ::Ptr-ified pointer.

Thanks. Easier now to go through the other changes. Yea, some extra lines for using that but didn't seem too bad.

pvuorela commented:
Also indeed the Extended... classes seem somewhat pointless

dcaliste commented:
Thank you for the review. Besides addressing your remarks, I've added tests in every routines with a ::Ptr argument to guard against null() values. I removed them when removing the ::Ptr types but I forgot to put them back...

pvuorela reviewed (approved):
Looking good. Nicely cleans up the handler code too.
